processing/src/main/java/org/apache/carbondata/processing/newflow/steps/writer/DataWriterProcessorStepImpl.java

```diff
@@ -0,0 +1,360 @@
+      } catch (Exception e) {
+        LOGGER.error(e, "Failed for table: " + tableName + " in  finishing data handler");
+      }
+    }
+    LOGGER.info("Record Procerssed For table: " + tableName);
+    String logMessage =
+        "Finished Carbon DataWriterProcessorStepImpl: Read: " + readCounter + ": Write: "
+            + writeCounter;
+    LOGGER.info(logMessage);
+    CarbonTimeStatisticsFactory.getLoadStatisticsInstance().recordTotalRecords(writeCounter);
+    processingComplete();
+    CarbonTimeStatisticsFactory.getLoadStatisticsInstance()
+        .recordDictionaryValue2MdkAdd2FileTime(configuration.getPartitionId(),
+            System.currentTimeMillis());
+    CarbonTimeStatisticsFactory.getLoadStatisticsInstance()
+        .recordMdkGenerateTotalTime(configuration.getPartitionId(), System.currentTimeMillis());
+
+    return null;
+  }
+
+  @Override public void close() {
+
+  }
+
+  private void processingComplete() throws CarbonDataLoadingException {
+    if (null != dataHandler) {
+      try {
+        dataHandler.closeHandler();
+      } catch (CarbonDataWriterException e) {
+        LOGGER.error(e, e.getMessage());
+        throw new CarbonDataLoadingException(e.getMessage());
+      } catch (Exception e) {
+        LOGGER.error(e, e.getMessage());
+        throw new CarbonDataLoadingException("There is an unexpected error: " + e.getMessage());
+      }
+    }
+  }
+
+  private void processBatch(CarbonRowBatch batch) throws CarbonDataLoadingException {
+    Iterator<CarbonRow> iterator = batch.getBatchIterator();
+    try {
+      while (iterator.hasNext()) {
+        CarbonRow row = iterator.next();
+        readCounter++;
+        Object[] outputRow = null;
+        // adding one for the high cardinality dims byte array.
+        if (noDictionaryCount > 0 || complexDimensionCount > 0) {
+          outputRow = new Object[measureCount + 1 + 1];
+        } else {
+          outputRow = new Object[measureCount + 1];
+        }
+
+        int l = 0;
+        int index = 0;
+        Object[] measures = row.getObjectArray(measureIndex);
+        for (int i = 0; i < measureCount; i++) {
+          outputRow[l++] = measures[index++];
+        }
+        outputRow[l] = row.getBinary(noDimByteArrayIndex);
+
+        int[] highCardExcludedRows = new int[segmentProperties.getDimColumnsCardinality().length];
+        Integer[] dimsArray = row.getIntegerArray(dimsArrayIndex);
+        for (int i = 0; i < highCardExcludedRows.length; i++) {
+          highCardExcludedRows[i] = dimsArray[i];
+        }
+
+        outputRow[outputRow.length - 1] = keyGenerator.generateKey(highCardExcludedRows);
+        dataHandler.addDataToStore(outputRow);
+        writeCounter++;
+      }
+    } catch (Exception e) {
+      throw new CarbonDataLoadingException("unable to generate the mdkey", e);
+    }
+  }
+
+  @Override protected CarbonRow processRow(CarbonRow row) {
+    return null;
```

Comment:
if this is returning null, in `AbstractDataLoadProcessorStep.processRowBatch` it is unnecessarily adding many null